File: educational-core/src/com/jetbrains/edu/coursecreator/actions/sections/CCCreateSection.java
package com.jetbrains.edu.coursecreator.actions.sections;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.Function;
import com.jetbrains.edu.coursecreator.CCUtils;
import com.jetbrains.edu.coursecreator.actions.CCCreateStudyItemActionBase;
import com.jetbrains.edu.coursecreator.actions.NewStudyItemInfo;
import com.jetbrains.edu.coursecreator.actions.StudyItemType;
import com.jetbrains.edu.coursecreator.actions.StudyItemVariant;
import com.jetbrains.edu.learning.OpenApiExtKt;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.Section;
import com.jetbrains.edu.learning.courseFormat.StudyItem;
import com.jetbrains.edu.learning.messages.EduCoreBundle;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;

import static icons.EducationalCoreIcons.Section;

public class CCCreateSection extends CCCreateStudyItemActionBase<Section> {

  public CCCreateSection() {
    super(StudyItemType.SECTION, Section);
  }

  @Override
  protected void addItem(@NotNull Course course, @NotNull Section section) {
    course.addSection(section);
  }

  @NotNull
  @Override
  protected Function<VirtualFile, ? extends StudyItem> getStudyOrderable(@NotNull final StudyItem item,
                                                                         @NotNull Course course) {
    return file -> course.getItem(file.getName());
  }

  @Override
  @Nullable
  protected VirtualFile createItemDir(@NotNull final Project project, @NotNull final Course course, @NotNull final Section section,
                                      @NotNull final VirtualFile parentDirectory) {
    return CCUtils.createSectionDir(project, section.getName());
  }

  @Override
  protected int getSiblingsSize(@NotNull Course course, @Nullable StudyItem parentItem) {
    return course.getItems().size();
  }

  @Nullable
  @Override
  protected StudyItem getParentItem(@NotNull Project project, @NotNull Course course, @NotNull VirtualFile directory) {
    return course;
  }

  @Nullable
  @Override
  protected StudyItem getThresholdItem(@NotNull Project project, @NotNull final Course course, @NotNull final VirtualFile sourceDirectory) {
    return course.getItem(sourceDirectory.getName());
  }

  @Override
  protected boolean isAddedAsLast(@NotNull Project project, @NotNull Course course, @NotNull VirtualFile sourceDirectory) {
    return sourceDirectory.equals(OpenApiExtKt.getCourseDir(project));
  }

  @Override
  protected void sortSiblings(@NotNull Course course, @Nullable StudyItem parentItem) {
    course.sortItems();
  }

  @Override
  protected void initItem(@NotNull Project project,
                          @NotNull Course course,
                          @Nullable StudyItem parentItem,
                          @NotNull Section item,
                          @NotNull NewStudyItemInfo info) {
    item.setCourse(course);
  }

  @NotNull
  @Override
  protected List<StudyItemVariant> getStudyItemVariants() {
    return Collections.singletonList(
      new StudyItemVariant(StringUtil.toTitleCase(EduCoreBundle.message("study.item.section")), "", Section, Section::new)
    );
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/actions/studyItem/CCCreateSection.kt
package com.jetbrains.edu.coursecreator.actions.studyItem

import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.util.Function
import com.jetbrains.edu.coursecreator.StudyItemType
import com.jetbrains.edu.coursecreator.presentableTitleName
import com.jetbrains.edu.learning.courseDir
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.courseFormat.Section
import com.jetbrains.edu.learning.courseFormat.StudyItem
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils
import icons.EducationalCoreIcons
import java.io.IOException

class CCCreateSection : CCCreateStudyItemActionBase<Section>(StudyItemType.SECTION_TYPE, EducationalCoreIcons.Section) {
  override fun addItem(course: Course, item: Section) {
    course.addSection(item)
  }

  override fun getStudyOrderable(item: StudyItem, course: Course): Function<VirtualFile, out StudyItem?> {
    return Function<VirtualFile, StudyItem> { file: VirtualFile -> course.getItem(file.name) }
  }

  @Throws(IOException::class)
  override fun createItemDir(project: Project, course: Course, item: Section, parentDirectory: VirtualFile): VirtualFile? {
    return GeneratorUtils.createSection(item, parentDirectory)
  }

  override fun getSiblingsSize(course: Course, parentItem: StudyItem?): Int = course.items.size

  override fun getParentItem(project: Project, course: Course, directory: VirtualFile): StudyItem? = course

  override fun getThresholdItem(project: Project, course: Course, sourceDirectory: VirtualFile): StudyItem? {
    return course.getItem(sourceDirectory.name)
  }

  override fun isAddedAsLast(project: Project, course: Course, sourceDirectory: VirtualFile): Boolean {
    return sourceDirectory == project.courseDir
  }

  override fun sortSiblings(course: Course, parentItem: StudyItem?) {
    course.sortItems()
  }

  override fun initItem(project: Project, course: Course, parentItem: StudyItem?, item: Section, info: NewStudyItemInfo) {
    item.course = course
  }

  override val studyItemVariants: List<StudyItemVariant>
    get() = listOf(
      StudyItemVariant(StudyItemType.SECTION_TYPE.presentableTitleName, "", EducationalCoreIcons.Section) { Section() }
    )
}
